src/kakarot/kakarot_proxy.cairo

```diff
@@ -0,0 +1,177 @@
+    implementation: felt
+) {
+    let (implementation) = Proxy.get_implementation_hash();
+    return (implementation,);
+}
+
+// @dev Cairo doesn't support native decoding like Solidity yet,
+//      that's why we pass three arguments for calldata instead of one
+// @param implementation_hash the implementation contract hash
+// @param selector the implementation initializer function selector
+// @param calldata_len the calldata length for the initializer
+// @param calldata an array of felt containing the raw calldata
+@constructor
+func constructor{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    implementation_hash: felt, selector: felt, calldata_len: felt, calldata: felt*
+) {
+    alloc_locals;
+    AccessControl._set_role_admin(IMPLEMENTATION, ADMIN);
+    // calldata[0] == owner
+    AccessControl._grant_role(ADMIN, calldata[0]);
+    AccessControl._grant_role(IMPLEMENTATION, calldata[0]);
+
+    Proxy._set_implementation_hash(implementation_hash);
+
+    if (selector != 0) {
+        // Initialize proxy from implementation
+        library_call(
+            class_hash=implementation_hash,
+            function_selector=selector,
+            calldata_size=calldata_len,
+            calldata=calldata,
+        );
+    }
+
+    return ();
+}
+
+//
+// Access Control functions
+//
+
+// @notice Allows the admin to transfer the admin role
+// @param user - Address of the new admin
+@external
+func transfer_admin_role{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    user: felt
+) {
+    alloc_locals;
+    AccessControl.assert_only_role(ADMIN);
+    let (local caller) = get_caller_address();
+    AccessControl._grant_role(ADMIN, user);
+    AccessControl._revoke_role(ADMIN, caller);
+    return ();
+}
+
+// @notice Grant the implementation role to a given address
+// @param user - Address of the user who will receive the implementation role
+@external
+func grant_implementation_role{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    user
+) {
+    AccessControl.assert_only_role(ADMIN);
+    AccessControl._grant_role(IMPLEMENTATION, user);
+    return ();
+}
+
+// @notice revoke the implementation role from a given address
+// @param user - Address of the user will have the implementation role revoked
+@external
+func revoke_implementation_role{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    user
+) {
+    AccessControl.assert_only_role(ADMIN);
+    AccessControl._revoke_role(IMPLEMENTATION, user);
+    return ();
+}
+
+//
+// Implementation Control
+//
+
+// @notice set the implementation contract class hash for this proxy
+// @param _new_implementation - The class hash of the implementation
+@external
+func set_implementation_hash{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+    _new_implementation: felt
+) {
+    AccessControl.assert_only_role(IMPLEMENTATION);
```

Comment:
what does this mean?
only the implementation role can change the implementation?

Reply:
Yes exactly.
The OZ admin setup is that you have one role that has admin permissions to change other roles.
And then additional roles for other functionality.
The ideas was that some one like you could have the admin role and then give the Implementation role to other people that can then change the kakarot proxy implementation.